Make observing LiveData more readable

lotdrops opened the pull request:
I do not like how LiveData is observed and I've also found a lot of people using `this` as lifecycleOwner in fragments. 
I did these extension functions to solve the first and reduce the likelihood of the second.
If you like them the same mechanism could be applied to your EventSource.observe.

Zhuinden commented on app/src/main/java/com/zhuinden/jetpacknavigationdaggersavedstatehandleftueexperiment/utils/LiveDataUtils.kt:
```diff
@@ -15,22 +15,74 @@
         override fun onChanged(t: T) {
-            removeObserver(this)
-            onChanged.invoke(t)
+            liveData.removeObserver(this)
```
nice catch

Zhuinden commented:
The idea of hiding the `viewLifecycleOwner` is sound, but I do not like the approach.

I have 1 nitpick and 1 real concern about it.

``` kotlin
/**
 * Utils to observe liveData in a cleaner way from any LifecycleOwner.
 *
 * This should not be used with Fragment as lifecycleOwner
 */
inline fun <T> LifecycleOwner.observeLifeCycle(
    liveData: LiveData<T>,
    crossinline observer: (T) -> Unit
) {
    liveData.observe(this, Observer { observer(it) })
}
```

Nitpick: it's called `Lifecycle` in the lifecycle api, not `LifeCycle`, this can easily throw off users of the code.

Real issue: if this shouldn't be called on Fragment, then I should not have an extension function that makes it possible (in this case, `LifecycleOwner.observe()`).

Adding 1 extension for a AppCompatActivity, and 1 extension for a Fragment is sufficient, and more importantly safer to use.

----

This also shows that I should remove `observeOnce()` because it is not actually used, it is left-over from a previous attempt when I was trying to initialize the ViewModel from `onCreate()` (without luck, initially moving it to `onCreateView`, then through lazy access to `onViewCreated`).

--------

I'm slightly conflicted about adding this though because making the `viewLifecycleOwner` explicit in the sample might help people know that that's what they should be using, especially considering that [even the Google Dev Advocates tend to screw this up](https://github.com/google/iosched/blob/4054aa3f8934b8b1208d5823fdbf531a8eb367af/mobile/src/main/java/com/google/samples/apps/iosched/ui/sessiondetail/SessionDetailFragment.kt#L217).

lotdrops commented:
True that it should not be LifeCycle, my bad.

In my case I have the comments mainly to educate about using the viewLifecycleOwner, and I left them to not loose that in case someone didn't realize why there are separate methods for fragment and activity. They could be removed if not helpful, or changed for a link to a good resource about the fragment lifecycle.
The LifecycleOwner extension could be made private. I left it for the rare occasions where you are not in a fragment or activity and you have a lifecycleOwner to use. But in those cases it is probably better to use the official api and reduce the risk of people using the wrong lifecycle in fragment.
And I saw you were not using the observeOnce. I thought you left it to use soon so I added the equivalent (I sometimes use it myself). If it won't be used it can be removed.

About what is better, whether to leave the viewLifecycleOwner to be more explicit or have it in a utils, it's hard to tell. In my case I preferred the second one but I also had doubts about that... My reasoning was that people who care about details would notice it with or without the extension, and people who do not would not research it enough so they would probably forget about it. Therefore, I preferred readability, hiding slightly lower level details, and making mistakes easier to detect in a code review (I think).

I fixed the nitpick and real issue. Whether it should be added or not is up to you :) Maybe I gave you more trouble than help with this PR

Zhuinden commented:
This won't be merged. Thanks for the PR though 😊

est7 commented:
`inline fun <T> LifecycleOwner.observe(
    liveData: LiveData<T>,
    crossinline observer: T.() -> Unit
) = liveData.observe(this, Observer { observer(it) })
`
is it correct
